activerecord/lib/active_record/connection_adapters/abstract/query_comment.rb

```diff
@@ -0,0 +1,143 @@
+
+require "active_support/core_ext/module/attribute_accessors_per_thread"
+
+module ActiveRecord
+  module ConnectionAdapters
+    # SQL query comments set at the adapter level
+    module QueryComment
+      mattr_accessor :prepend_comment, default: false
+      def add_query_comment?
+        return false unless ActiveRecord::Base.query_comment_enabled
+        return false if Context.components.empty?
+        true
+      end
+
+      def add_query_comment_to_sql(sql)
+        return sql unless add_query_comment?
+        Context.set(adapter: self)
+        comment = Context.comment
+        if comment.present? && !sql.include?(comment)
+          sql = if prepend_comment
+            "#{comment} #{sql}"
+          else
+            "#{sql} #{comment}"
+          end
+        end
+        sql
+      end
+
+      # Maintains a user-defined context for all queries and constructs an SQL comment by
+      # calling methods listed in :components
+      #
+      # Additional information can be pushed to the context to be referenced by methods
+      # defined in framework or application initializers.
+      module Context
+
+        mattr_accessor :components, :lines_to_ignore
+        mattr_accessor :cache_query_comment, default: false
+        thread_mattr_accessor :context, default: {}
+        thread_mattr_accessor :cached_comment
+        DEFAULT_LINES_TO_IGNORE_REGEX = %r{\.rvm|/ruby/gems/|vendor/|active_record|active_support|rbenv|monitor\.rb.*mon_synchronize}
+
+        class << self
+          Context.components = [:application]
+
+          def comment
+            return cached_comment if cache_query_comment && !cached_comment.nil?
+            self.cached_comment = "/*#{escape_sql_comment(comment_content)}*/"
+          end
+
+          def escape_sql_comment(str)
+            while str.include?('/*') || str.include?('*/')
+              str = str.gsub('/*', '').gsub('*/', '')
+            end
+            str
+          end
+
+          def comment_content
+            components.filter_map do |c|
+              if component_value = self.try(c)
+                "#{c}:#{component_value}"
+              end
+            end.join(",")
+          end
+
+          def set(ctx)
+            return unless ctx.is_a? Hash
+            self.context.merge!(ctx)
+            self.cached_comment = nil
+          end
+
+          def context_available?(context_name)
+            context.key?(context_name) && context[context_name].present?
+          end
+
+          def clear_comment_cache!
+            self.cached_comment = nil
+          end
+
+          # Default ActiveRecord comment components
+          # :application, :connection_config, :line
+          def application
+            if defined?(Rails.application)
+              context[:application_name] ||= Rails.application.class.name.split("::").first
+            else
+              context[:application_name] ||= "rails"
+            end
+            context[:application_name]
```

Comment:
I'm not entirely sure but I think that this can be set by the Active Record Railties initializer (in activerecord/lib/active_record/railtie.rb) on boot - if that's possible then we wouldn't need to check for Rails.application - since the Railtie has access.